### ISSUE TITLE
Can not resolve path properly for resources defined over variables

### ISSUE THREAD

**Issue by Grygir** [bholloway/resolve-url-loader#126]
**resolve-url-loader** in the version 3.1.0, has been changed the way how references from sass variables are resolved.

When sass variable contains relative reference to resource file, that reference got resolved from the place where the variable is used, not from the place where the variable is defined.
```
|- src
   |- images
      |- a.png
   |- styles
      |- mobile
         |- main.scss
      |- _variabls.scss
      |- main.scss
```
**src/styles/_variabls.scss**
```scss
$background: url('../images/a.png') !default;
```

**src/styles/main.scss**
```scss
@import "variables";
.head {
    background: $background;
}
```

And **src/styles/mobile/main.scss**
```scss
@import "../variables";
.head {
    background: $background;
}
```

Leads to the error
```
ERROR in ./src/styles/mobile/main.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleNotFoundError: Module not found: Error: Can't resolve '../images/a.png' in '/src/styles/mobile'
```

Note. In version 2.3.1 resources from variables are resolved fine.

**Comment by bholloway** [comment]
Please check verbose output with  debug option.

The variables never source mapped correctly. In v2 there was possible a
full file search was happening which is non performant.

Search is no longer supported in v3 but you can write a custom join to do
whatever you need.

**Comment by Grygir** [comment]
Have checked with increased verbosity level, but don't see any information that could help to solve the problem.

config
```js
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

module.exports = {
    mode: 'development',
    entry: {
        style: './src/styles/feature.scss',
        'mobile-style': './src/styles/mobile/feature.scss'
    },
    output: {
        publicPath: '/dist/',
        filename: '[name].js'
    },
    module: {
        rules: [
            {
                test: /\.s?css$/,
                use: [{
                    loader: MiniCssExtractPlugin.loader,
                }, {
                    loader: 'css-loader',
                    options: {}
                }, {
                    loader: 'resolve-url-loader',
                    options: {}
                }, {
                    loader: 'sass-loader',
                    options: {
                        debug: true,
                        sourceMap: true,
                        sourceMapContents: false
                    }
                }]
            },
            {
                test: /\.png$/,
                loader: 'url-loader',
                options: {
                    limit: 1,
                    emitFile: false,
                    publicPath: '/',
                    name: '[path][name].[ext]?[hash]'
                }
            }
        ]
    },
    plugins: [
        new MiniCssExtractPlugin({
            filename: '[name].css'
        })
    ]
};
```
command
```sh
webpack --config webpack.config.js --display verbose --verbose --debug
```
output
```
    ERROR in ./src/styles/mobile/main.scss (./node_modules/css-loader/dist/cjs.js??ref--4-1!./node_modules/resolve-url-loader??ref--4-2!./node_modules/sass-loader/lib/loader.js??ref--4-3!./src/styles/mobile/main.scss)
    Module not found: Error: Can't resolve '../images/a.png' in './src/styles/mobile'
    resolve '../images/a.png' in './src/styles/mobile'
      using description file: ./package.json (relative path: ./src/styles/mobile)
        Field 'browser' doesn't contain a valid alias configuration
        using description file: ./package.json (relative path: ./src/styles/images/a.png)
          no extension
            Field 'browser' doesn't contain a valid alias configuration
            ./src/styles/images/a.png doesn't exist
          .wasm
            Field 'browser' doesn't contain a valid alias configuration
            ./src/styles/images/a.png.wasm doesn't exist
          .mjs
            Field 'browser' doesn't contain a valid alias configuration
            ./src/styles/images/a.png.mjs doesn't exist
          .js
            Field 'browser' doesn't contain a valid alias configuration
            ./src/styles/images/a.png.js doesn't exist
          .json
            Field 'browser' doesn't contain a valid alias configuration
            ./src/styles/images/a.png.json doesn't exist
          as directory
            ./src/styles/images/a.png doesn't exist
    [./src/styles/images/a.png]
    [./src/styles/images/a.png.wasm]
    [./src/styles/images/a.png.mjs]
    [./src/styles/images/a.png.js]
    [./src/styles/images/a.png.json]
     @ ./src/styles/mobile/main.scss (./node_modules/css-loader/dist/cjs.js??ref--4-1!./node_modules/resolve-url-loader??ref--4-2!./node_modules/sass-loader/lib/loader.js??ref--4-3!./src/styles/mobile/main.scss) 4:38-64
```

**Comment by bholloway** [comment]
As a side note, if you have all your assets in an `images` directory then you might prefer `postcss-url` instead. This loader helps you locate assets with your sass partials in a feature-based file structure.

**Comment by Grygir** [comment]
Sorry, my mistake, have added `debug` option to the wrong loader.
Adding to the `'resolve-url-loader'`
```js
               {
                    loader: 'resolve-url-loader',
                    options: {
                        debug: true
                    }
                }
```
extends the output with following error trace
```
ERROR in ./src/styles/mobile/feature.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleNotFoundError: Module not found: Error: Can't resolve '../images/a.png' in './src/styles/mobile'
    at ./node_modules/webpack/lib/Compilation.js:821:10
    at ./node_modules/webpack/lib/NormalModuleFactory.js:401:22
    at ./node_modules/webpack/lib/NormalModuleFactory.js:130:21
    at ./node_modules/webpack/lib/NormalModuleFactory.js:224:22
    at ./node_modules/neo-async/async.js:2830:7
    at ./node_modules/neo-async/async.js:6877:13
    at ./node_modules/webpack/lib/NormalModuleFactory.js:214:25
    at ./node_modules/enhanced-resolve/lib/Resolver.js:184:12
    at ./node_modules/enhanced-resolve/lib/Resolver.js:238:5
    at eval (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
    at ./node_modules/enhanced-resolve/lib/UnsafeCachePlugin.js:37:5
    at ./node_modules/enhanced-resolve/lib/Resolver.js:238:5
    at eval (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
    at ./node_modules/enhanced-resolve/lib/Resolver.js:238:5
    at eval (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:27:1)
    at ./node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:42:38
    at ./node_modules/enhanced-resolve/lib/Resolver.js:238:5
    at eval (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)
    at ./node_modules/enhanced-resolve/lib/Resolver.js:238:5
    at eval (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:27:1)
    at ./node_modules/enhanced-resolve/lib/DescriptionFilePlugin.js:42:38
    at ./node_modules/enhanced-resolve/lib/Resolver.js:238:5
    at eval (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)
    at ./node_modules/enhanced-resolve/lib/Resolver.js:238:5
    at eval (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:15:1)
    at ./node_modules/enhanced-resolve/lib/DirectoryExistsPlugin.js:22:13
    at ./node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:73:15
    at processTicksAndRejections (internal/process/task_queues.js:75:11)
```

> if you have all your assets in an images directory then you might prefer postcss-url instead. 
Not sure, it will work on our case. We have code spread by bundles. Each bundle may have own assets with its images and styles with no strict requirements for directories names. But styles assembly is done with single webpack config.

**Comment by bholloway** [comment]
🤔Hmmm not the output I expect.

Is it possible you can make a github repo with minimum breaking example?

**Comment by Grygir** [comment]
Yes, sure. Here it is https://github.com/Grygir/resolve-url-loader-issue
Thanks.

**Comment by bholloway** [comment]
Sorry smashed with work hopefully in a couple of days I can take a look

**Comment by bholloway** [comment]
(Sorry for the delay)

I ran your breaking example as is and yes it is broken.
```
resolve-url-loader: ./src/styles/feature.scss: ../images/a.png
  ./src/styles
  FOUND
resolve-url-loader: ./src/styles/mobile/feature.scss: ../images/a.png
  ./src/styles/mobile
  NOT FOUND
```

Using version `^2.3.1` it works because of the file search "magic"
```
resolve-url-loader: ../images/a.png
  /Users/bholloway/Documents/bholloway/resolve-url-loader-issue/src/styles
  FOUND

resolve-url-loader: ../images/a.png
  /Users/bholloway/Documents/bholloway/resolve-url-loader-issue/src/styles/mobile
  /Users/bholloway/Documents/bholloway/resolve-url-loader-issue/src/styles
  FOUND
```

The file search was removed in v3 because it was not deterministic and not performant.

🤔 I will think on this and play with some code before commenting further.

**Comment by bholloway** [comment]
Some analysis...

In your project directory I ran
```
>  node-sass src/styles/mobile/feature.scss --source-map-embed > compiled.css
```

I then went to [sokra's sourcemap visualisation page](http://sokra.github.io/source-map-visualization/#custom) and uploaded the `compiled.css` file and the requested source files.

Here is the result
- brown is `mobile/feature.scss`
- blue is `_variables.scss`

![image](https://user-images.githubusercontent.com/6868309/61949286-aad97a00-afed-11e9-9066-bdedd839e9e0.png)

Adding `console.log(declaration.source.start);` to the source-code [here](https://github.com/bholloway/resolve-url-loader/blob/master/packages/resolve-url-loader/lib/engine/postcss.js#L63) the result is.

```
{ line: 2, column: 3 }
```

Meaning
- the correct info is in the sourcemap to make this work correctly for you
- `resolve-url-loader` is looking at the position of the declaration (brown) and not the value (blue)

Experiment...

I played further with postcss. There is no sourcemap location for the value (blue) only the whole statement. However we can use `declaration.prop` (e.g. `background`) and `declaration.raws.between` (e.g. `: `) to find the offset to the value.

I designed v3 with the option to iterate a number of possible candidate directories in order to find a match. So it is possible we can pass both locations to the join operation. I tested this and it works.

Next steps...

I'll need to construct some test cases, particularly with multiline declarations.

I can then amend the `postcss` engine to check first the value position and then the declaration position.

**Comment by gabsprates** [comment]
same problem here :confused:

**Comment by jcamilleri13** [comment]
I've just bumped into the error too. Is there an active PR we could help contribute to perhaps? Or maybe help with test cases? I'm trying to find a workaround for now but it would be great if we can correct the behaviour. Thanks!

**Comment by bholloway** [comment]
I had a basic proof of concept but I have not progressed it.

Let me dig up the code and see how much would take to release a simple @next version for you to try.

**Comment by jcamilleri13** [comment]
That would be great, thanks a ton!

**Comment by bholloway** [comment]
This is non-trivial but I do have something in progress.

**Comment by bholloway** [comment]
update: I have an implementation that passes existing tests, I just need to finalise a test case for this particular issue.

Once I do that I'll have a better idea how to release, as `distag next` or as `engine: "next"` in a minor release.

**Comment by bholloway** [comment]
*Input Needed*

What is our search order out of these..
1. The actual position of the uri inside the `url(...)` statement
2. The start of the `url()` statement
3. The start of the declaration value (in `property: value;`)
4. The start of the declaration property (in `property: value;`)

Noting
* In the previous `engine: postcss` it was just `4`.
* Right now I have implemented `[4,3,1]`.

I feel it is most correct as `[1,2,3,4]`. Thoughts?

**Comment by jcamilleri13** [comment]
Good idea to have it search multiple locations, I didn't think of that. It'll allow for more flexible project structures, although it should probably be explicitly mentioned in the documentation so it doesn't just work "by magic" and it'll be easier to debug if things go wrong. 

Regarding the first option, I presume it can handle string concatenation and interpolation, right? That's something sass itself should take care of I suppose.


Your suggested ordering seems to mimic the way things would actually be defined within a sass code base (i.e. You can't really have the url part defined before its contents, and the property existing without a value) so it seems like a sensible approach. I actually have a project structure that would allow a user to override a file with an identically named one that's "closer" to the source declaration, that ordering would tackle situations like that by working "outwards" from the uri declaration.

I'd be happy to help test on Windows if you'd like. Thanks for looking into this!

**Comment by bholloway** [comment]
So I'm hearing from @jcamilleri13 that the order `[1,2,3,4]` seems okay(?).

> "Regarding the first option, I presume it can handle string concatenation and interpolation, right? That's something sass itself should take care of I suppose."

Yes this all happens after sass (or whatever) has done its job. Its looking postfactum at the source-map. So if you're asking if we have to work around variables interpolating in sass then no that is not an issue.

I can query the source-map at any text position so it is possible to do `1`. Interestingly `postcss` actually doesn't have resolution to that level, only to the `url()` statement. But it is easy to do `1` anyway for possible future use.

Whereas `2` is possible and should be useful with multiple `url()` statements in a declaration with different source locations. Its only in writing this up did I realise I didn't do it. I will need to implement this next.

> "I actually have a project structure that would allow a user to override a file with an identically named one that's "closer" to the source declaration"

Yes! this is exactly what I'm worried about.

This structure potentially makes resolution feel non-deterministic. **@jcamilleri13 can you please talk more about your project structure and your motivations.**

> "I'd be happy to help test on Windows if you'd like."

That will be good. I run the e2e tests on MacOs and windows but in terms of _actual projects_ I do zero for any platform.

Right now its looking like I will keep the two `engine: rework|postcss` as is and just do a `@next` and then `4.0.0` release for the new postcss functionality. It is just too hard to support the known-breaking e2e test cases otherwise.

**Comment by jcamilleri13** [comment]
Hey @bholloway,

So regarding my project, in a nutshell it's something like this. I work on a pretty large application that is resold to our clients. This application is generally themed/skinned to match the client's brand, so we override sass variables containing colours, logo paths, and other such configuration to generate their specific stylesheets. The scenario I'm envisaging would be something like this:
```
\images
  background.jpg
/styles
  /global
  /themes
    /theme1
     variables.scss *
    /theme2
      /images
       background.jpg
    variables.scss **
```

In the first theme, the variable for the login background is referencing the global/default background image, but the client using theme two would like a custom background so that is loaded into their theme folder and referenced from there. I would expect the customised file (**) to load the "adjacent" `background.jpg` but the un-tweaked one in theme one (*) to load the more external image file. Not sure if I've made myself clear enough.

The search order you proposed should handle this use case, as the root would be the correct theme file I'm loading the image for (**).

**Comment by jcamilleri13** [comment]
Quick question... would this help solve the issue about how mixins are handled (or rather not handled), or will that issue be unaffected by this update?

**Comment by bholloway** [comment]
@jcamilleri13 I'm glad you reminded me about this.

**Problem**

Having a look it appears the sourcemap resolves to the `@mixin` definition. This is because we walk the declarations of the AST and the declaration is created by the mixin regardless of where it is included. This is consistent with my statement in the "known issues".

🤔I suspect that I was using "mixin" to mean mixins where you pass variables.
```
@mixin something($url) {
  background-image: url($url);
}
```
```
.some-selector {
  @include something("image.jpg");
}
```
4 years ago when this started I was really keen on this for some reason I've since forgotten. Probably sprite sheets or something. Anyway I'm confident it is something to do with this case.

In this case you would want to resolve the image to the second file, at the point of use/invocation where you are writing the url. However the sourcemap will point to the mixin definition.

**Solution**

It is trivial for me to add the `declaration.parent` in the current changes. Which should fix this problem.

So right now I have implemented:
1. ~The actual position of the uri inside the url(...) statement~
2. The start of the url() statement
3. The start of the declaration value (in property: value;)
4. The start of the declaration property (in property: value;)
5. The start of the selector (or specifically the declaration parent)

I will finalise testing and publish an v4 alpha in a few days 🤞

**Comment by jcamilleri13** [comment]
So by adding the `declaration.parent` to the list, the search will also check the file where the mixin was called as a last resort? Brilliant. This will handle two issues at once. Thanks for putting the time into this, really appreciate it. As soon as the alpha's published I'll let it loose on my project on Windows and WSL and let you know how it handles the use cases.

**Comment by bholloway** [comment]
Published as `resolve-url-loader@4.0.0-alpha.1`

**Comment by jcamilleri13** [comment]
So I've started testing this in the project, I'll update here if I encounter anything. So far, it seems like the mixins still seem top be causing problems, but I'm not sure the use-case I've tested so far is being handled. If the url() statement is actually created in the mixin (not passed in as a parameter), is the new engine expected to trace up to the parent and test there, or will this only take place if the url was created externally and passed in as a parameter?

**Comment by bholloway** [comment]
@jcamilleri13 it will _want_ to resolve where the `url()` is but it will _eventually_ resolve to the file where the selector is. So if you have that project structure with the overrides it will match on the`@mixin` definition.

**Comment by ILokalin** [comment]
Hello.
Pleas, is version 4 already available?
I have a similar problem. "resolve-url-loader": "^3.1.0"

**Comment by bholloway** [comment]
Only the alpha @ILokalin. It’s on dist tag ‘next’.

**Comment by chrisguitarguy** [comment]
I think this is impacting bootstrap 3 icons + sass-loader. Downgrading to 2.3.X works fine, however.

```scss
$icon-font-path: "~bootstrap-sass/assets/fonts/";
@import "~bootstrap-sass/assets/stylesheets/bootstrap";
```

**Comment by angelicochris** [comment]
Just want to confirm that the alpha indeed works for this where 3.1.1 did not.

**Comment by jlhogan** [comment]
does anyone have a workaround here?

**Comment by bholloway** [comment]
@jlhogan please open a new issue with your specific use case and I will try to help

**Comment by bholloway** [comment]
Released `resolve-url-loader@4.0.0-beta.1`.

Since `v4` is now `master` branch and `v3` has moved to maintenance so I'm going to close this issue.

If you cannot adopt the beta in your project, or `v4` full release is not released in the next few weeks, then feel free to reopen this issue.
